modules/bq-log-alerting/logging/cloud_function/index.js

```diff
@@ -0,0 +1,165 @@
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+const {BigQuery} = require('@google-cloud/bigquery');
+const {SecurityCenterClient} = require('@google-cloud/security-center')
+const MD5 = require('crypto-js/md5');
+
+const bigquery = new BigQuery();
+// These constants will need to be tuned to your preference
+const project = process.env.LOGGING_PROJECT;
+const dataset_name = 'views';
+const invocation_name = 'securityalert_invocation';
+const result_name = 'securityalert_result';
+const source_name = process.env.CSCC_SOURCE;
+const dry_run = (process.env.DRY_RUN === 'true');
+
+// Not currently used, was for original version of the cloud function
+getMetric = function(name, labels) {
+  return {
+    type: `custom.googleapis.com/${name}`,
+    labels: labels
+  };
+}
+
+getResource = function(labels) {
+  return {
+    type: 'global',
+    labels: labels
+  };
+}
+
+getTimeseries = function(value) {
+  return [{
+    interval: {endTime: {seconds: Date.now() / 1000}},
+    value: value
+  }];
+}
+
+function createFindingObject (source_name, labels) {
+  const eventTime = new Date(labels.eventTimestamp.value).getTime();
+  const sourceProperties = {};
+  for (const key in labels) {
+    if (key !== 'eventTimestamp' &&
+        key !== 'resourceName' &&
+        key !== 'query') {
+      sourceProperties[key] = { stringValue: "" + labels[key] };
+    }
+  }
+  return {
+    parent: "" + source_name,
+    findingId: MD5("" + labels.query + labels.eventTimestamp + labels.callerIp + labels.principalEmail + labels.resourceName),
+    finding: {
+      state: 'ACTIVE',
+      // Resource the finding is associated with.  This is an
+      // example any resource identifier can be used.
+      resourceName: labels.resourceName,
+      // A free-form category.
+      category: "" + labels.query,
+      // The time associated with discovering the issue.
+      eventTime: {
+        seconds: Math.floor(eventTime / 1000),
+        nanos: (eventTime % 1000) * 1e6,
+      },
+      sourceProperties: sourceProperties
+    }
+  }
+}
+
+async function createFinding(cscc_client, source_name, labels) {
+  try {
+    finding = createFindingObject(source_name, labels)
+    if (dry_run) {
+      console.log('DRY_RUN: scc finding: ', JSON.stringify(finding));
+    } else {
+      const [newFinding] = await cscc_client.createFinding(finding);
+    }
+    return 1;
+  } catch (err) {
+    console.error("" + err);
+  }
+  return 0;
+}
+
+async function createFindingsFromResult(cscc_client, source_name, tableName, result) {
```

Comment:
I am very much a JS newbie so excuse my lack of understanding, but can we do this in parallel as `createFinding` is async, so map and wait for promises to resolve?

Reply:
I think we can the sync version first and try the async version later.